### ISSUE TITLE
Plot will be failed with "plot without title" after a few hours intact.

### ISSUE THREAD

**Issue by andrewzhang1** [jupyter/notebook#1806]
I use latest anacoda python 2.7 on ubunt (64 bits) and installed R (conda install -c r r-essentials). Everything worked fine for a few hours for both ipyton and R (calcalation and visualizations); however, after a few hours, I will find faied to re-run any of the plot with R code (ipython notebood is still fine). 

For example: 

1) This code will be run fine with plot after I just start or re-start the ipython notebook:

library(ggplot2)
small <- diamonds[sample(nrow(diamonds), 1000),]
ggplot(small) + geom_point(aes(x=carat, y=price)) + ggtitle('Example 1: Diamonds: Price over Crate')

However, it will be failed after a few hours if I re-run it and get a message:
"plot without title"

2) This code will be run fine with plot after I just start or re-start the ipython notebook:
# Define 2 vectors
cars <- c(1, 3, 6, 4, 9)
trucks <- c(2, 5, 4, 5, 12)

# Graph cars using a y axis that ranges from 0 to 12
plot(cars, type="o", col="blue", ylim=c(0,12))

# Graph trucks with red dashed line and square points
lines(trucks, type="o", pch=22, lty=2, col="red")

# Create a title with a red, bold/italic font
title(main="Autos", col.main="red", font.main=4)

However, it will be failed after a few hours if I re-run it and get a message:
Plot with title “Autos”

Even I restart the R kernel, it still failed; I have to kill the current ipython notebook process and re-start the ipython notebook, then it will be working again. After a few hours intact, it will fail again. 

This is how I start the ipython notebook in the background: 

nohup ipython notebook   --ip='*' --port=9000 > start_ipython_notebook_anaconda2-by-user-conda2-port-9000.log &

Not sure what caused this failure.

**Comment by takluyver** [comment]
Pinging @flying-sheep for an issue related to R.

It sounds like after some time it's deciding to only generate the text representation of the plot. I don't think I've ever seen that before, and I can't think why it would occur.

**Comment by flying-sheep** [comment]
has to be notebook-related, not R or R kernel related:

> Even if I restart the R kernel, it will still be failed; I have to kill the current ipython notebook process and re-start the ipython notebook, then it will be working fine again. After a few hours intact, it will fail again.

**Comment by takluyver** [comment]
Right, it can't be R in isolation, but this is so weird I wondered if you knew of anything that might cause the kernel to behave oddly while plotting.

@andrewzhang1 when this happens, can you open your browser's JS console and see if there are any errors displayed there?

**Comment by flying-sheep** [comment]
so it’s certainly the notebook process that goes bad somehow.

he only thing a notebook process gone bad could influence the kernel would be if it after that point repeatedly sends a message to the kernel that re-breaks the kernel even after restarts… e.g. if it starts to prefix every code cell with `options(jupyter.plot_mimetypes = NULL)` or repeatedly sends that code to be executed. this sounds implausible.

therefore i guess the problem is probably completely on the notebook side, although it *might* be only possible with the kind of data the IRkernel sends (certain PNG structure? no idea)

@andrewzhang1 could you please also send us a notebook that was saved with one intact plot cell (before the damage, displaying a plot) and a broken plot cell (after the damage, displaying the “Plot with(out) title” text)

**Comment by andrewzhang1** [comment]
[Sample_R_Before-Damage.ipynb.zip](https://github.com/jupyter/notebook/files/509223/Sample_R_Before-Damage.ipynb.zip)
[Sample_R_After-Damage.ipynb.zip](https://github.com/jupyter/notebook/files/509224/Sample_R_After-Damage.ipynb.zip)

I enclosed two ipynb files: Sample_R_Before-Damage.ipynb and Sample_R_After-Damage.ipynb. Please let know if I need to add more, thanks!

**Comment by andrewzhang1** [comment]
I checd the version for the python and R in my two machines that have above problem:
R version 3.3.1 (2016-06-21) -- "Bug in Your Hair"

Python 2.7.12 |Anaconda 2.2.0 (64-bit)| (default, Jul  2 2016, 17:42:40)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2

and 

R version 3.3.1 (2016-06-21) -- "Bug in Your Hair"

Python 2.7.12 |Anaconda 4.1.1 (64-bit)| (default, Jul  2 2016, 17:42:40)

**Comment by andrewzhang1** [comment]
However, I had an old instance with same os (ubuntu) on aws that runs for many months, and the both R and ipytyon notebooks work fine without above problem, their version of R and python are:

R version 3.2.2 (2015-08-14) -- "Fire Safety"

Python 2.7.11 |Anaconda 2.4.1 (64-bit)| (default, Dec  6 2015, 18:08:32)

Is the problem due to the newer version of R?

**Comment by flying-sheep** [comment]
hmm, so the PNG data is certainly gone from the file… i.e. not a pure display problem.

and as said: i don’t think R has anything to do with it, since as you said the corruption persists across kernel restarts but not notebook restarts.

**Comment by takluyver** [comment]
Are there any errors in the JS console when you try to plot something and the plot doesn't show?

**Comment by andrewzhang1** [comment]
@takluyver yes, sometime I saw this error after a few days intact when I try to plot:

E
r
r
o
r
 
i
n
 
.
E
x
t
e
r
n
a
l
2
(
C
_
X
1
1
,
 
p
a
s
t
e
(
"
p
n
g
:
:
"
,
 
f
i
l
e
n
a
m
e
,
 
s
e
p
 
=
 
"
"
)
,
 
g
$
w
i
d
t
h
,
 
:
 
u
n
a
b
l
e
 
t
o
 
s
t
a
r
t
 
d
e
v
i
c
e
 
P
N
G


T
r
a
c
e
b
a
c
k
:



plot without title

**Comment by flying-sheep** [comment]
That old error with the exploded string again!

```py
In [6]: ''.join(c if c else ' ' for c in s.split('\n'))
Out[6]: 'Error in .External2(C_X11, paste("png::", filename, sep ""), g$width, : unable to start device PNG Traceback:'
```

**Comment by takluyver** [comment]
So that is an error coming from R. I'm mystified as to why it would start doing that depending on how long the notebook server has been running, though. :-S

**Comment by flying-sheep** [comment]
wild ass guess: maybe they get never closed and just open file descriptors until the OS tells them not to?

and the file descriptors somehow get repatriated to the jupyter process once the kernel dies and therefore don’t vanish?

IDK why i think this, but maybe it’s better to mention it than not.

**Comment by takluyver** [comment]
Reasonable hypothesis, but to the best of my knowledge file descriptors are cleaned up when a process dies.

@andrewzhang1 are you totally sure that restarting an R kernel after this has occurred does not make it work again? Sorry to ask something that you already said, but that seems very strange. Is there anything else that could be changing when the problem appears?

**Comment by andrewzhang1** [comment]
Not it's even worse: it starts hanging when I run any of the R code, it will show: 
In [*]; while ipython notebook works fine for the plot, etc. Even I restart the R kernel, it won't solve the problem, I have to kill the process and re-start the ipyton notebook in order to get it back to work (I saw this many times, and use it as a workaround by re-starting the notebook).

**Comment by flying-sheep** [comment]
maybe an update gone awry. your error message is mentioned here:

http://stackoverflow.com/questions/22096391/generate-png-with-r

**Comment by andrewzhang1** [comment]
@flying-sheep I installed the update as per above links; however, it did no solve my problem. It still get the error after a while. What is the root cause for the problem? What if I downgrade to R 3.2 (how do I downgrade)?

**Comment by flying-sheep** [comment]
we’re in the dark about the root cause, i’m afraid.

the error above (“unable to start device PNG”) reminds me of another error that occurs when R isn’t installed properly, but then it would appear immediately, not after a time.

the question is if you or your sysadmin compiled R from source. if so, it will compile but have broken graphics (i.e. `png` doesn’t work). you can fix that by

1. using R from package sources instead of compiling yourself, or
2. by recompiling with something like `./configure --prefix=/usr/local --without-x --with-cairo --with-libpng --enable-R-shlib && make && make install`, and then checking if `getOption('bitmapType') == 'cairo`

**Comment by andrewzhang1** [comment]
@lchutny I recently did not encounter any issues since I upgraded to latest conda: 
https://anaconda.org/anaconda/anaconda
https://anaconda.org/r/r-irkernel

Guess it's fixed. Good luck!
